test/utils/wabt_engine.cpp

```diff
@@ -37,123 +49,128 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
-            results[0].set_i32(ret);
-            return wabt::interp::Result::Ok;
+            results[0].Set(ret);
+            return wabt::Result::Ok;
         });
 
-    wabt::Errors errors;
-    const wabt::Result result = wabt::ReadBinaryInterp(&m_env, wasm_binary.data(),
-        wasm_binary.size(), wabt::ReadBinaryOptions{}, &errors, &m_module);
-    if (result != wabt::Result::Ok)
-        return false;
-    assert(m_module != nullptr);
+    wabt::interp::Trap::Ptr trap;
+    m_instance =
+        wabt::interp::Instance::Instantiate(m_store, m_module.ref(), {host_func->self()}, &trap);
 
-    // Run start function (this will return ok if no start function is present)
-    const wabt::interp::ExecResult r = m_executor.RunStartFunction(m_module);
-    return r.result == wabt::interp::Result::Ok;
+    return m_instance && !trap;
 }
 
 bool WabtEngine::init_memory(bytes_view memory)
 {
-    if (m_env.GetMemoryCount() == 0)
-        return false;
+    if (!m_instance || m_instance->memories().empty())
+        return {};
 
-    auto& dst = *m_env.GetMemory(0);
-    if (dst.data.size() < memory.size())
+    wabt::interp::Memory::Ptr dst_memory(m_store, m_instance->memories()[0]);
+    if (dst_memory->ByteSize() < memory.size())
         return false;
 
-    std::memcpy(dst.data.data(), memory.data(), memory.size());
+    std::memcpy(dst_memory->UnsafeData(), memory.data(), memory.size());
     return true;
 }
 
 bytes_view WabtEngine::get_memory() const
 {
-    if (m_env.GetMemoryCount() == 0)
+    if (!m_instance || m_instance->memories().empty())
         return {};
 
-    auto& env = const_cast<wabt::interp::Environment&>(m_env);
-    const auto& memory = env.GetMemory(0);
-    return {reinterpret_cast<uint8_t*>(memory->data.data()), memory->data.size()};
+    wabt::interp::Memory::Ptr memory(m_store, m_instance->memories()[0]);
+    return {memory->UnsafeData(), memory->ByteSize()};
 }
 
 std::optional<WasmEngine::FuncRef> WabtEngine::find_function(
     std::string_view name, std::string_view) const
 {
-    const wabt::interp::Export* e = m_module->GetExport({name.data(), name.size()});
-    if (e != nullptr && e->kind == wabt::ExternalKind::Func)
-        return reinterpret_cast<WasmEngine::FuncRef>(e);
-    return {};
+    if (!m_module)
+        return {};
+
+    const auto it_export = std::find_if(m_module->desc().exports.begin(),
+        m_module->desc().exports.end(), [name](const wabt::interp::ExportDesc& e) {
+            return (e.type.type->kind == wabt::ExternalKind::Func && e.type.name == name);
+        });
+    if (it_export == m_module->desc().exports.end())
+        return {};
+
+    return WasmEngine::FuncRef{it_export->index};
 }
 
 WasmEngine::Result WabtEngine::execute(
     WasmEngine::FuncRef func_ref, const std::vector<uint64_t>& args)
 {
-    const auto* e = reinterpret_cast<const wabt::interp::Export*>(func_ref);
+    assert(m_instance);
```

Comment:
Why assert here and not in init_memory/get_memory? I think we have the "rules" described in `wasm_engine.h`, I can't remember myself.

Reply:
`wasm_engine.hpp` says for all three`Requires instantiate().`, though it seems ambiguous whether it means "assumes instantiate was called" or "fails when instantiate was not called".

But I see that here it's inconsitent now, I guess I'll change it to asserts in all three, `fizzy_engine` also assumes  `m_instance` is non-null there.